src/libraries/System.Text.Json/src/System/Text/Json/Serialization/DefaultReferenceResolver.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the.NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+
+namespace System.Text.Json
+{
+    internal sealed class DefaultReferenceResolver
+    {
+        private uint _referenceCount;
+        private Dictionary<string, object>? _keyObjectMap;
+        private Dictionary<object, string>? _objectKeyMap;
+
+        public DefaultReferenceResolver(bool isWrite)
+        {
+            if (isWrite)
+            {
+                _objectKeyMap = new Dictionary<object, string>(ReferenceEqualsEqualityComparer<object>.Comparer);
+            }
+            else
+            {
+                _keyObjectMap = new Dictionary<string, object>();
+            }
+        }
+
+        public void AddReference(string key, object value)
+        {
+            if (_keyObjectMap!.ContainsKey(key))
```

Comment:
The TryAdd method is faster since it doesn't require the upfront check. See `JsonHelpers.TryAdd` which wraps that to also support netfx.